geniza/corpus/management/commands/add_links.py

```diff
@@ -255,32 +272,82 @@ def add_link(self, row):
         # get doc relation for this link type
         doc_relation = self.document_relation[link_type]
 
-        # get source based on the link type
+        # get source and location based on the link type
         if link_type == "goitein_note":
             source = self.get_goitein_source(doc=doc, title="typed texts")
+            base_name = basename(row["link_target"])
+            try:  # use base name up to PGPID for goitein texts
+                end_index = base_name.index("(PGPID")
+            except ValueError:  # if PGPID not present, use base name up to extension
+                end_index = base_name.rindex(".") if "." in base_name else 0
+            # use string up to end_index and strip whitespace for location
+            location = base_name[:end_index].strip() if end_index else base_name.strip()
         elif link_type == "indexcard":
             source = self.get_goitein_source(
                 doc=doc,
                 title="index cards",
-                url="https://geniza.princeton.edu/indexcards/",
             )
+            # use card #target for location
+            location = "card #%s" % row["link_target"]
         elif link_type == "jewish-traders":
             source = self.jewish_traders
+            # use document number without extension (e.g. "01.pdf" becomes "01") for location
+            base_name = row["link_target"]
+            no_ext = (
+                base_name[: base_name.rindex(".")] if "." in base_name else base_name
+            )
+            m = re.search(r"(\d+)", no_ext)
+            location = ("document #%s" % m.groups(0)) if m else no_ext
         elif link_type == "india-traders":
             source = self.get_india_book(row["link_title"])
+            # use filename without extension (e.g. "I-27-28.pdf" becomes "I-27-28") for location
+            base_name = row["link_target"]
+            no_ext = (
+                base_name[: base_name.rindex(".")] if "." in base_name else base_name
+            )
+            location = no_ext.strip()
 
         # create or update footnote with the target url, beasd on all the other parameters determined
         return self.set_footnote_url(
-            doc=doc, source=source, url=target_url, doc_relation=doc_relation
+            doc=doc,
+            source=source,
+            url=target_url,
+            doc_relation=doc_relation,
+            location=location,
+        )
+
+    def generate_csv_report(self, csv_path, csv_headers):
+        """Generate a CSV containing the rows of the original CSV for which a document
+        could not be found."""
+
+        # basename without extension
+        original_csv = (
+            basename(csv_path)[: basename(csv_path).rindex(".")]
+            if "." in csv_path
+            else basename(csv_path)
+        )
+        # save in same directory as original CSV
+        nf_report_path = dirname(csv_path) + (
+            "/documents_not_found_in_%s.csv" % original_csv
         )
```

Comment:
needs to handle input csv file in current working directory! (as I found when I ran it locally)

how about this?

(possibly we should use `os.path.join` while we're at it...)

```suggestion
        nf_report_path = "%s/documents_not_found_in_%s.csv" % (dirname(csv_path)  or '.', original_csv)
```